osquery/events/windows/usn_journal_reader.cpp

```diff
@@ -0,0 +1,879 @@
+  // been a problem yet, but we're not doing that check + correction below.
+  while (current_buffer_ptr < buffer_end_ptr) {
+    const auto current_record =
+        reinterpret_cast<const USN_RECORD*>(current_buffer_ptr);
+
+    const auto next_buffer_ptr =
+        current_buffer_ptr + current_record->RecordLength;
+    if (next_buffer_ptr > buffer_end_ptr) {
+      return Status(1, "Received a malformed USN_RECORD. Terminating...");
+    }
+
+    auto status =
+        ProcessAndAppendUSNRecord(record_list,
+                                  current_record,
+                                  d_->per_file_last_record_type_map,
+                                  d_->journal_reader_context->drive_letter);
+    if (!status.ok()) {
+      LOG(ERROR) << status.getMessage();
+    }
+
+    current_buffer_ptr = next_buffer_ptr;
+  }
+
+  return Status(0);
+}
+
+void USNJournalReader::dispatchEventRecords(
+    const std::vector<USNJournalEventRecord>& record_list) {
+  if (record_list.empty()) {
+    return;
+  }
+
+  // Notify the publisher that we have new records ready to be acquired
+  auto& context = d_->journal_reader_context;
+
+  {
+    WriteLock lock(context->processed_records_mutex);
+
+    context->processed_record_list.reserve(
+        context->processed_record_list.size() + record_list.size());
+
+    context->processed_record_list.insert(context->processed_record_list.end(),
+                                          record_list.begin(),
+                                          record_list.end());
+
+    context->processed_records_cv.notify_all();
+  }
+}
+
+void USNJournalReader::start() {
+  // Acquire a handle to the device, as well as the journal id and the initial
+  // sequence number
+  auto status = initialize();
+  if (!status.ok()) {
+    LOG(ERROR) << status.getMessage();
+    return;
+  }
+
+  // Enter the main loop, listening for journal changes
+  while (!interrupted() && !d_->journal_reader_context->terminate) {
+    status = acquireRecords();
+    if (!status.ok()) {
+      LOG(ERROR) << status.getMessage();
+      return;
+    }
+
+    // Process the records
+    std::vector<USNJournalEventRecord> record_list;
+    status = processAcquiredRecords(record_list);
+    if (!status.ok()) {
+      LOG(ERROR) << status.getMessage();
+      return;
+    }
+
+    // Send the new records to the event publisher
+    dispatchEventRecords(record_list);
+  }
+}
+
+void USNJournalReader::stop() {
+  if (d_->volume_handle != INVALID_HANDLE_VALUE) {
+    ::CloseHandle(d_->volume_handle);
+    d_->volume_handle = INVALID_HANDLE_VALUE;
+  }
+}
+
+USNJournalReader::USNJournalReader(
+    USNJournalReaderContextRef journal_reader_context)
+    : InternalRunnable("USNJournalReader"), d_(new PrivateData) {
+  d_->journal_reader_context = journal_reader_context;
+}
+
+USNJournalReader::~USNJournalReader() {}
+
+// TODO: Needs tests; see https://github.com/osquery/osquery/issues/5847
+Status USNJournalReader::DecompressRecord(
```

Comment:
Need more comments here.  It seems like the purpose of this is to deduplicate (not decompress)?  Removing consecutive events for the same file with the same 'reason' flag?

Reply:
> Removing consecutive events for the same file with the same 'reason' flag?

Not quite -- from the `USN_RECORD` docs:

> Reason
The flags that identify reasons for changes that have accumulated in this file or directory journal record since the file or directory opened.
When a file or directory closes, then a final USN record is generated with the USN_REASON_CLOSE flag set. The next change (for example, after the next open operation or deletion) starts a new record with a new set of reason flags.
A rename or move operation generates two USN records, one that records the old parent directory for the item, and one that records a new parent.

The change journal saves space/minimizes record creation by combining all but the last (i.e. close) operation into a single record with different bits for different actions set high. For example, open + read + write + truncate + change attributes + close results in only two records: one containing the first five actions, and the last containing just the close action. `DecompressRecord` just iterates over the bits that we know (from the documentation) and makes a copy of the record containing each reason independently. I can add that as a comment, if you'd like.